### DISCUSSION TITLE
v1.3.1

### DISCUSSION THREAD

**Discussion by txubelaxu** [txubelaxu/esphome-jk-bms#15] [Announcements]
Some bugs corrected.
Performance improvements.

<hr /><em>This discussion was created from the release <a href='https://github.com/txubelaxu/esphome-jk-bms/releases/tag/v1.3.1'>v1.3.1</a>.</em>

**Comment by x230-thinkpad** [comment]
hi, can you help me with connection flickering
![image](https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/4ddc12d1-817f-44d8-9f80-d97d2f0317bf)

[logs_esp-rs485-gw_logs.txt](https://github.com/user-attachments/files/16056143/logs_esp-rs485-gw_logs.txt)

**Reply by x230-thinkpad** [reply]
> hi, can you help me with connection flickering ![image](https://private-user-images.githubusercontent.com/159946521/344740478-4ddc12d1-817f-44d8-9f80-d97d2f0317bf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk4NDk3NDQsIm5iZiI6MTcxOTg0OTQ0NCwicGF0aCI6Ii8xNTk5NDY1MjEvMzQ0NzQwNDc4LTRkZGMxMmQxLTgxN2YtNDRkOC05ZjgwLWQ5N2QyZjAzMTdiZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzAxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcwMVQxNTU3MjRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zOTQ0ODBkZmI2YzczNzdkMjlhMzZlODc2MzhiN2UxZDY5ZmRiZDA1NjJlMGQ1YTlkZjk1ZjI5MmJkYjY5M2IzJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.ZkSuVqpaajcndKaoTq1w1tTIogp88NnIrsrX9WF_VFk)
> 
> [logs_esp-rs485-gw_logs.txt](https://github.com/user-attachments/files/16056143/logs_esp-rs485-gw_logs.txt)

here I attach yaml file
[esp-rs485-gw.zip](https://github.com/user-attachments/files/16056162/esp-rs485-gw.zip)

**Comment by txubelaxu** [comment]
Sorry for deleting your comment at v1.3

If you want to work in mode2 you should not use address 0x00 for the first BMS.
The master uses that address. In mode2 ESP32 is the master, so you can use 0x01 for first BMS and 0x02 for second one.

You can try the example: esp32-S3-example-jkpb-rs485_0master_2slaves.yaml

Somebody recomended to me:
+ disable web_server lines
+ do not use INFO log level. Use DEBUG

**Reply by Patrick197030100** [reply]
Hi thank you for everything you help us! well I am in Victron protocol system on bms 0 master and 01 slave I cannot put the master bms on 0x01 because the Victron system no longer knows who the master is and it no longer adds the capacity of the 2 batteries so I have to leave it on the dip 000000 address 0x00 on the master that's the problem!!!

@Imanol82

**Comment by Patrick197030100** [comment]
hello is ok look tank you  look 
![Screenshot 2024-07-01 at 18-41-54 jkbms 1et2 – Home Assistant](https://github.com/txubelaxu/esphome-jk-bms/assets/148679768/81c0c71a-a5af-4802-8efd-bdc5eca0f3c4)
look the identification no good ?

**Comment by x230-thinkpad** [comment]
> Sorry for deleting your comment at v1.3
> 
> 
> 
> If you want to work in mode2 you should not use address 0x00 for the first BMS.
> 
> The master uses that address. In mode2 ESP32 is the master, so you can use 0x01 for first BMS and 0x02 for second one.
> 
> 
> 
> You can try the example: esp32-S3-example-jkpb-rs485_0master_2slaves.yaml
> 
> 
> 
> Somebody recomended to me:
> 
> + disable web_server lines
> 
> + do not use INFO log level. Use DEBUG

looks good, many thanks mate. almost 30 minutes keep stable connection

![image](https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/33916c6d-1cd6-45c4-8df1-507a55d972ab)

**Reply by Patrick197030100** [reply]
HELLO Yes I understood but on the master if I put Address Dip 01 for Master and 02 for slave the Victron Protocol walks more because he will see that only one BMS in // and not 2 on JKBMS Victron Le Master protocol and always 0x00 here issue !!!!

**Comment by Patrick197030100** [comment]
hello no good for bms master 0x01 and slave 0x02 !

**Reply by txubelaxu** [reply]
Please, use DEBUG level. Post here the log.txt, please.

**Comment by x230-thinkpad** [comment]
after around 1 hour nonstop connect, the ESP is stop responding meanwhile status still online,
 I have removed the API contain and add static dedicated ip for ESP32, pls help there is something wrong with my yaml 

![image](https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/5d7b872b-0d33-4c43-9300-8f764d7ddfe0)

[logs_esp-rs485-gw_logs.txt](https://github.com/user-attachments/files/16065004/logs_esp-rs485-gw_logs.txt)
[esp-rs485-gw_v1.3.0.zip](https://github.com/user-attachments/files/16065023/esp-rs485-gw_v1.3.0.zip)


many thx.

**Reply by txubelaxu** [reply]
I think your RX/TX pinout is not correct. Could you test with 1.3.1 version, please?
Use esp32-example-jkpb-rs485_0master_2slaves.yaml as the base for your config, ok?

**Reply by x230-thinkpad** [reply]
> I think your RX/TX pinout is not correct. Could you test with 1.3.1 version, please? Use esp32-example-jkpb-rs485_0master_2slaves.yaml as the base for your config, ok?

hi mate, 

RX-TX pin correct already, I have follow esp32-example-jkpb-rs485_0master_2slaves.yaml
but it still flickering connection

![image](https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/ae460f25-b39f-4ac5-843d-182d4f625041)

![image](https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/fbde3147-4224-45ee-bc75-0fab7449360e)

I think, if I put wrong pin it should be not appear the value at all

![image](https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/5d34ced2-55de-4261-a3ed-87e29689377b)

**Reply by txubelaxu** [reply]
Could you post, please, your complete last esp-rs485-gw.yaml and the complete log?

Thanks

**Reply by x230-thinkpad** [reply]
> Could you post, please, your complete last esp-rs485-gw.yaml and the complete log?
> 
> Thanks

here I attached the yaml file mate,
[esp-rs485-gw-Last.zip](https://github.com/user-attachments/files/16076174/esp-rs485-gw-Last.zip)

logs: 
API Encryption Enable : [logs_esp-rs485-gw_run.txt](https://github.com/user-attachments/files/16076295/logs_esp-rs485-gw_run.txt)
API Encryption Disabled : [logs_esp-rs485-gw_run_v2.txt](https://github.com/user-attachments/files/16076394/logs_esp-rs485-gw_run_v2.txt)

at the end, the logs is stuck like this
![image](https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/26637f5d-2477-4c33-b4c0-ee4155bd0474)

I used JK-PB 15.XA hardware version, and 15.23 software version
![image](https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/75e3c898-3dd5-49ef-9049-79fc80ffc685)

Thank You

**Reply by txubelaxu** [reply]
Hi @x230-thinkpad,

The information arrives via UART is not correct:
[10:43:28][D][jk_rs485_sniffer:311]: Buffer before number 1:    FF.FF.F8.FF.FF.EF.FF.FF.FF.FE.FF.FC.FF.FF.7F.FF.FF.FF.FF.FF.FF.FF.FF.FF.FE.BE.FC.7C.FF.FE.E8.FF.FE.FC.FF.FF.FB.FF.DF.FF.FF.FF.FF.FF.F7.FF.7C.FB.FF.FC.FF.FC.FF.FF.FF.EF.7F.FC.FB.FF.FF.7E.F0.FF.FF.FF.FF.FF.FF.FF.F7.FF.7C.FE.FF.FF.FF.FF.FF.FF.FF.EF.FF.FF.FF.FF.FF.FF.00.FF.FF.FF.EF.FF.B8.FE.FF.EF.EF.1F.F8.FF.FC.FE.FF.F7.FF.FF.FE.F8.FF.F8.EF.FF.FE.F8.FF.E0.FF.FF.FF.FC.EF.FF.E0.FF.FF.FF.FF.FF.FF.FE.FF.FE.FF.FF.FF.FE.E0.FF.FF.FF.FF.7D.FF.FE.EF.FF.FF.FF.

The problem is the information that is detected in de UART. 
Something wrong in the serial or the RS485 network. Have you conected the GND wire in RS485 network (3 wires in total)

**Reply by x230-thinkpad** [reply]
> ed the GND wire in R

I only put 2 wires from RS485 to UART TTL, 
A = orange
B = white orange 
VCC to 3v ESP32
GDN to GND ESP32

![WhatsApp Image 2024-07-03 at 19 00 10](https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/78c4e869-2f0f-4828-aefb-aae583f3d3c9)

should I plug the RS485 ground cable into the UART ?

**Comment by Patrick197030100** [comment]
ok thank you very good ok

2 juillet 2024 18:26 "x230-thinkpad" ***@***.*** ***@***.******@***.***>)> a écrit:
	I think your RX/TX pinout is not correct. Could you test with 1.3.1 version, please? Use esp32-example-jkpb-rs485_0master_2slaves.yaml as the base for your config, ok?

	hi mate,

	RX-TX pin correct already, I have follow esp32-example-jkpb-rs485_0master_2slaves.yaml
but it still flickering connection

	image.png (view on web) (https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/ae460f25-b39f-4ac5-843d-182d4f625041)

	image.png (view on web) (https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/fbde3147-4224-45ee-bc75-0fab7449360e)

	I think, if I put wrong pin it should be not appear the value at all

	image.png (view on web) (https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/5d34ced2-55de-4261-a3ed-87e29689377b)

**Comment by Patrick197030100** [comment]
is no good the yaml esp32-example-jkpb-rs485_0master_2slaves.yaml ? look picture

2 juillet 2024 19:16 "Txubelaxu" ***@***.*** ***@***.******@***.***>)> a écrit:
	Could you post, please, your complete last esp-rs485-gw.yaml and the complete log?

	Thanks

**Comment by txubelaxu** [comment]
connect GND from RS485 wire to ESP32 GND ad well

**Reply by x230-thinkpad** [reply]
> connect GND from RS485 wire to ESP32 GND ad well El El mié, 3 jul 2024 a las 13:03, x230-thinkpad ***@***.***> escribió:
> […](#)
> ed the GND wire in R I only put 2 wires from RS485 to UART TTL, A = orange B = white orange VCC to 3v ESP32 GDN to GND ESP32 WhatsApp.Image.2024-07-03.at.19.00.10.jpeg (view on web) <https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/78c4e869-2f0f-4828-aefb-aae583f3d3c9> — Reply to this email directly, view it on GitHub <[#15 (reply in thread)](https://github.com/txubelaxu/esphome-jk-bms/discussions/15#discussioncomment-9946878)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/BFHIJMCR3JEGLRLR7CSVU43ZKPLABAVCNFSM6AAAAABKFY3GXOVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TSNBWHA3TQ> . You are receiving this because you authored the thread.Message ID: ***@***.***>

I will try, but is there any issue with components mate? can't detected external_components due to index file corrupted

![image](https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/d28bf6c1-bd36-4b63-9a91-5831c8144b5a)

**Reply by txubelaxu** [reply]
I have compiled your yaml with only 2 mods:
1st: uncomment line: components: [ jk_rs485_sniffer , jk_rs485_bms ]
2nd: comment two lines below "ota:"

Maybe you have to launch "esphome clean ....." before "esphome compile ....."

**Reply by x230-thinkpad** [reply]
> I have compiled your yaml with only 2 mods: 1st: uncomment line: components: [ jk_rs485_sniffer , jk_rs485_bms ] 2nd: comment two lines below "ota:"
> 
> Maybe you have to launch "esphome clean ....." before "esphome compile ....."

looks good, many thanks mate

![image](https://github.com/txubelaxu/esphome-jk-bms/assets/159946521/b8df8bc4-7ed3-4159-9e92-1f8edf10acd5)

**Reply by txubelaxu** [reply]
Congratulations!!!

**Comment by Imanol82** [comment]
Arratsaldeon.

First, thank you for the work you have done. You make it very easy for those of us who don't know.
I have used the yaml "esp32-example-jkpb-rs485_1master_1slave.yaml". I have 2 BMSs in parallel.

I have 2 questions. 

1. Why is the BMS INFO only seen on the slave? Have I done something wrong?

- Master:
![image](https://github.com/txubelaxu/esphome-jk-bms/assets/109697268/fa34670e-20ed-44ab-a1b4-bb37e4112dd0)

- Slave:
![image](https://github.com/txubelaxu/esphome-jk-bms/assets/109697268/57192c6a-9969-4fa0-96c6-68b66c8498b1)

2. And a question that I'm even embarrassed to ask but I haven't been able to do it. How do I update to version 1.31?

Thank you very much.

Mila esker! :)

**Reply by txubelaxu** [reply]
Hi Imanol,

Answers:

1. In a RS485 network can only be ONE master. So, in your case, ESP32 cannot act as master. Why? because there is a master yet in your network (one of your BMSs). The problem is that BMS device INFO only is answered by an slave to a master request to it. So, nobody is going to ask to your master. My advice is:
   1) to configure your BMS both as slave (for example: address 0x01 and 0x02)
   2) use 0master_2slaves… yaml file and modify to your needs.
   3) when you get this working, you can try esp-jk-bms-can proyect. There is a development branch named “development” that is using txubelaxu’s code. This code adds a CAN support for your inverter. So, BMSs—>ESP32–>INVERTER (via CAN BUS)
2. Well, I am very newbie at github. You could download the ZIP package from that version and compile, run, upload…

Tell me please, your advances…

Thanks!!

**Reply by Imanol82** [reply]
But I have already CAN comunication between the Master BMS and Inverter. 
You are suggested not to use this CAN comunication, to put the ESP as Master and comunicate with CAN the ESP and the Inverter?

Thank you very much!

**Reply by txubelaxu** [reply]
This second option (esp-jk-bms-can) has some advantages, in my opinion. You know what are the rules the system uses and you can set/monitor via HA as well. Take a look at https://github.com/Sleeper85/esphome-jk-bms-can 
The problem is that you need a ESP2CAN converter.

**Reply by Imanol82** [reply]
I don't know if I'm convinced that an ESP and HA control the communication. Right now I have the MASTER BMS communicated with the DEYE inverter and it is more or less going well.

If I put the 2 BMSs as slaves and leave them communicating via CAN with the inverter, will they stop working?

![image](https://github.com/txubelaxu/esphome-jk-bms/assets/109697268/91a1cb2b-79e8-4c17-9ae6-85bc54187e36)

Eskerrikasko!

**Reply by Imanol82** [reply]
I answer myself:
I have set the 2 BMSs as slave and everything seems to be going well.

1. The inverter continues to see the information.
![image](https://github.com/txubelaxu/esphome-jk-bms/assets/109697268/6085430f-d5ab-40b3-953b-d9dd1871544d)

2. Now all the information is working out for me.
![image](https://github.com/txubelaxu/esphome-jk-bms/assets/109697268/5a5a85de-5f58-452a-9120-085d9b7b0082)
![image](https://github.com/txubelaxu/esphome-jk-bms/assets/109697268/03c9c652-15c8-465e-a05b-112e15454f9d)


But I have a doubt, what SOC is seeing the Deye. Because 97% appears when one BMS is at 97% and the other at 100%.

Let's see how everything behaves.

Eskerrikasko!
Thank you very much

**Reply by Imanol82** [reply]
No, I answer myself again.

Putting the two as a slave, the Deye only sees one of the two. I have noticed because apart from the SOC, it only set a charge/discharge current of 133A instead of 266A.

So I've installed again 1Master 1Slave yaml.


Thank you vcery much

**Reply by txubelaxu** [reply]
https://github.com/Sleeper85/esphome-jk-bms-can/issues/58#issuecomment-2218361397

**Reply by Imanol82** [reply]
In the end I'll have to try it.  
Sissi's doesn't work for me, everything blocks.  
I don't really like that the ESP acts as master, if it fails, can it tell the inverter to increase the current to 5000A?  And do the BMS obey it?  Or is there simply the configuration (and protection) of each BMS?  But if the master changes any parameters?  
In the end I was satisfied with your RS485 project but with the real switches hehe, that would act on the BMS.  The rest, the JKBMSs-Investor communication is going very well indeed.

Eskerrikasko laguntzagatik!!

Thank you very much.

**Comment by txubelaxu** [comment]
can it tell the inverter to increase the current to 5000A?

No, it cannot. The BMS settings cannot be changed by ESP. I have not
developed for changing params. It is on my to do.

So, each BMS will respect your settings.

I do not know what the master BMS is setting on the slave BMS. I do not
know what could change on slave. But I would prefer to be under control.

**Reply by Imanol82** [reply]
OK understood.

I already told that I will try it. 
I have already bought the CAN converter and I already have the rest working, it's just adding the CAN.


Mila esker @txubelaxu .

Thank you very much

**Comment by Patrick197030100** [comment]
Hi thank you for everything you help us! well I am in Victron protocol system on bms 0 master and 01 slave I cannot put the master bms on 0x01 because the Victron system no longer knows who the master is and it no longer adds the capacity of the 2 batteries so I have to leave it on the dip 000000 address 0x00 on the master that's the problem!!!

@Imanol82

**Comment by Imanol82** [comment]
One question. This board is enough to do all?
Has RS485 and Can, no?

Acabo de encontrar este increíble artículo en AliExpress. ¡Échale un vistazo! 
27,55€ | T-CAN485 ESP32 CAN RS-485, compatible con tarjeta TF, WIFI, Bluetooth, módulo de Control inalámbrico IOT, placa de desarrollo
https://a.aliexpress.com/_EQcxhWr

**Comment by Imanol82** [comment]
Good afternoon.

I have no idea what I'm doing but something seems to come out. 
I have managed to configure the ESP up to this point. As soon as I can, I'll fix the RS485-ttl and the CAN. And let's see what happens.
I have used Mode 1 for my 2 JK BMS-PBs. I don't know if it's OK or not. 

I have also put the FULL packets so that all the information appears.

![image](https://github.com/user-attachments/assets/ca280118-6803-4ad5-becc-70657b928024)


If I leave everything connected as I have it now (BMS Master, Slave and connected to the Deye) and connect this ESP32 only through the RS485 part, what happens?
Can I ignore the "yambms" sensors, they won't do anything, and use the JK-BMS 1 and 2 yambms sensors and switches?
But keeping the comunication between BMSs and Deye?

Thank you very much @txubelaxu 


Ze eromena... ezdet ezer ulertzen... Eskerrikasko!!

**Comment by testpaul999** [comment]
@txubelaxu 
Hello, 
I have put this two lines

 uart1_protocol_number:
      name: "${bms_name} uart1 protocol number"
    uart2_protocol_number:
      name: "${bms_name} uart2 protocol number"  

Into my yaml file and flash my esp32 who I use with Bluetooth to the JK-BMS.
But I get not the right value back it  always shows „0“ as protocol.
And one question more can you pls. implement the CAN Protocol as well.

regards

**Comment by txubelaxu** [comment]
I do not understand your config. Do you mean that you are using BLE or
RS485?

About can protocol: I am using "esphome-jk-bms-can" from
https://github.com/Sleeper85/esphome-jk-bms-can
It uses "esphome-jk-bms" inside it.

**Reply by testpaul999** [reply]
Hello,

I use the BLE integration.
And my wish are read out and change the three protocols
uart1
uart2
CAN
At the moment I get only „0“ as protocol version for the both uart.
![IMG_4780](https://github.com/user-attachments/assets/bd61879c-9198-43f6-98f2-c06d6bcca55b)
And this is my yaml file
###################################
[Jk bms.yaml.txt](https://github.com/user-attachments/files/16508691/Jk.bms.yaml.txt)

**Comment by x230-thinkpad** [comment]
Hi, can you help me with this error message

[logs_jk-bms_compile.txt](https://github.com/user-attachments/files/16568380/logs_jk-bms_compile.txt)
